### PR TITLE
Optimized bilinear interpolation using TensorIterator

### PR DESCRIPTION
**Pull request by vfdev-5**
Related to #10482

Description:

- Optimized bilinear interpolation for 1d, 2d, 3d cases using TensorIterator

### Results

<details>
<summary>
Interpolation 2d - 6 thread(s)
</summary>

In | Out | Is contiguous | Channels last | master | this PR | speed-up
---|---|---|---|---|---|---
[1, 3, 320, 320] | [256, 256] | True | False | 0.3938 | 0.0782 | 5.0339
[1, 3, 320, 320] | [512, 512] | True | False | 1.5585 | 0.4105 | 3.7965
[1, 3, 320, 320] | [256, 256] | False | False | 0.3481 | 0.0760 | 4.5780
[1, 3, 320, 320] | [512, 512] | False | False | 1.5848 | 0.4091 | 3.8734
[1, 3, 320, 320] | [256, 256] | False | True | 1.2058 | 1.2034 | 1.0020
[1, 3, 320, 320] | [512, 512] | False | True | 4.8691 | 4.8537 | 1.0032
[32, 128, 64, 64] | [32, 32] | False | True | 6.3915 | 6.4041 | 0.9980
[32, 128, 64, 64] | [128, 128] | False | True | 166.1769 | 164.5621 | 1.0098
[32, 128, 64, 64] | [32, 32] | True | False | 3.7194 | 2.4720 | 1.5046
[32, 128, 64, 64] | [128, 128] | True | False | 86.6704 | 52.3754 | 1.6548
[1, 3, 500, 500] | [256, 256] | True | False | 0.3270 | 0.0792 | 4.1307
[1, 3, 500, 500] | [800, 800] | True | False | 3.3116 | 0.5567 | 5.9482
[1, 3, 500, 500] | [256, 256] | False | False | 0.3763 | 0.0773 | 4.8700
[1, 3, 500, 500] | [800, 800] | False | False | 3.2577 | 0.5590 | 5.8279

</details>

<details>
<summary>
Interpolation 1d - 6 thread(s)
</summary>

In | Out | Is contiguous | Channels last | master | this PR | speed-up
---|---|---|---|---|---|---
[4, 512, 320] | 256 | True | False | 0.2795 | 0.1032 | 2.7089
[4, 512, 320] | 512 | True | False | 0.5533 | 0.1888 | 2.9303

</details>

<details>
<summary>
Interpolation 3d - 6 thread(s)
</summary>

In | Out | Is contiguous | Channels last | master | this PR | speed-up
---|---|---|---|---|---|---
[1, 3, 16, 320, 320] | [8, 256, 256] | True | False | 4.4105 | 2.1236 | 2.0769
[1, 3, 16, 320, 320] | [32, 512, 512] | True | False | 83.9426 | 42.6641 | 1.9675
[1, 3, 16, 320, 320] | [8, 256, 256] | False | True | 15.5736 | 15.5758 | 0.9999
[1, 3, 16, 320, 320] | [32, 512, 512] | False | True | 272.4795 | 273.2745 | 0.9971

</details>

<details>
<summary>
Interpolation 2d - 1 thread(s)
</summary>

In | Out | Is contiguous | Channels last | master | this PR | speed-up
---|---|---|---|---|---|---
[1, 3, 320, 320] | [256, 256] | True | False | 1.0240 | 0.4145 | 2.4705
[1, 3, 320, 320] | [512, 512] | True | False | 4.0771 | 1.3836 | 2.9467
[1, 3, 320, 320] | [256, 256] | False | False | 0.9771 | 0.3270 | 2.9878
[1, 3, 320, 320] | [512, 512] | False | False | 4.1732 | 1.2209 | 3.4180
[1, 3, 320, 320] | [256, 256] | False | True | 1.5466 | 1.5363 | 1.0067
[1, 3, 320, 320] | [512, 512] | False | True | 6.1555 | 6.1199 | 1.0058
[32, 128, 64, 64] | [32, 32] | False | True | 27.6362 | 27.5901 | 1.0017
[32, 128, 64, 64] | [128, 128] | False | True | 468.6442 | 465.5163 | 1.0067
[32, 128, 64, 64] | [32, 32] | True | False | 20.1495 | 10.0694 | 2.0011
[32, 128, 64, 64] | [128, 128] | True | False | 400.0401 | 204.0662 | 1.9603
[1, 3, 500, 500] | [256, 256] | True | False | 0.8956 | 0.3366 | 2.6606
[1, 3, 500, 500] | [800, 800] | True | False | 8.6554 | 2.9530 | 2.9310
[1, 3, 500, 500] | [256, 256] | False | False | 1.0921 | 0.3385 | 3.2263
[1, 3, 500, 500] | [800, 800] | False | False | 8.9594 | 2.9627 | 3.0241

</details>

<details>
<summary>
Interpolation 1d - 1 thread(s)
</summary>

In | Out | Is contiguous | Channels last | master | this PR | speed-up
---|---|---|---|---|---|---
[4, 512, 320] | 256 | True | False | 1.5233 | 0.5027 | 3.0301
[4, 512, 320] | 512 | True | False | 3.0302 | 0.9735 | 3.1128

</details>

<details>
<summary>
Interpolation 3d - 1 thread(s)
</summary>

In | Out | Is contiguous | Channels last | master | this PR | speed-up
---|---|---|---|---|---|---
[1, 3, 16, 320, 320] | [8, 256, 256] | True | False | 12.0477 | 11.3196 | 1.0643
[1, 3, 16, 320, 320] | [32, 512, 512] | True | False | 222.8618 | 209.9955 | 1.0613
[1, 3, 16, 320, 320] | [8, 256, 256] | False | True | 17.9883 | 17.9937 | 0.9997
[1, 3, 16, 320, 320] | [32, 512, 512] | False | True | 380.7244 | 380.1916 | 1.0014

</details>

<details>
<summary>
Versions and build configs
</summary>

PyTorch master: 1.9.0.dev20210223
PyTorch master build setting:
```
BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=10.2, CUDNN_VERSION=7.6.5, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.9.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON,
```

PR : 1.9.0a0+74b172b
PR build setting:
```
BUILD_TYPE=Release, CUDA_VERSION=11.1, CUDNN_VERSION=8.0.5, CXX_COMPILER=/usr/bin/g++-7, CXX_FLAGS=-O3 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.9.0, USE_CUDA=1, USE_CUDNN=1, USE_EIGEN_FOR_BLAS=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=OFF, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=0, USE_OPENMP=ON,
```
</details>

This description is based on the benchmarks and the code from [here](https://github.com/vfdev-5/interpolate-tensoriterator/tree/master/step_six).

TL;DR
- Linear upsampling generic implementation using TensorIterator for Nd case (single loop function for 1d, 2d and 3d cases)
  - can be generalized to nearest, bicubic interpolation modes.
- works for channels first and last cases.
 
Joint work with Francisco Massa (@fmassa).

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
@fmassa let me know if you cannot review it. 

@vfdev-5 please benchmark non contiguous cases

**Comment by vfdev-5** [comment]
@VitalyFedyunin there is already channel last (non-contigous) case [here](https://github.com/vfdev-5/interpolate-tensoriterator/blob/master/step_six/pth_vs_PR_results.log.save#L54-L64).

Do you think about particular cases like `torch.rand(1, 3, 420, 420)[:, :, :320, :320]` ?

**Comment by fmassa** [comment]
@VitalyFedyunin I'll review this PR today

**Comment by vfdev-5** [comment]
@VitalyFedyunin as talked with @fmassa , i'd like to mention here a performance drawback with dispatch stub in our particular case for 3d linear interpolation.
I compare the execution times for between [my prototype code (used for development and as a codebase for this PR)](https://github.com/vfdev-5/interpolate-tensoriterator/tree/master/step_six) and PyTorch build with this PR:

```
---- Benchmark 3D ----
Input tensor: [1, 3, 16, 320, 320]
Num threads: 6

- Bench upsample_trilinear3d_cpu (2000 rounds) - downsampling to [8, 256, 256]
Elapsed time (ms): 2.11485
- Bench ti_upsample_trilinear3d_kernel_impl (2000 rounds) - downsampling to [8, 256, 256]
Elapsed time (ms): 1.05858

- Bench upsample_trilinear3d_cpu (2000 rounds) - upsampling to [32, 512, 512]
Elapsed time (ms): 44.3299
- Bench ti_upsample_trilinear3d_kernel_impl (2000 rounds) - upsampling to [32, 512, 512]
Elapsed time (ms): 26.0051
```

Expected result should have ~ similar times between `upsample_trilinear3d_cpu` and `ti_upsample_trilinear3d_kernel_impl` as it is more or less the same code.

I tried also to register manually the dispatch as here : https://github.com/pytorch/pytorch/compare/master...Quansight:upsample-tensor-iterator-another-dispatch, see aten/src/ATen/native/UpSampleTrilinear3d.cpp
```
REGISTER_ARCH_DISPATCH(upsample_trilinear3d_kernel, DEFAULT, &upsample_trilinear3d_kernel_impl);
REGISTER_AVX_DISPATCH(upsample_trilinear3d_kernel, &upsample_trilinear3d_kernel_impl);
REGISTER_AVX2_DISPATCH(upsample_trilinear3d_kernel, &upsample_trilinear3d_kernel_impl);
```
and looks like this way I could restore expected times
```
---- Benchmark 3D ----
Input tensor: [1, 3, 16, 320, 320]
Num threads: 6

- Bench upsample_trilinear3d_cpu (750 rounds) - downsampling to [8, 256, 256]
Elapsed time (ms): 1.05641
- Bench ti_upsample_trilinear3d_kernel_impl (750 rounds) - downsampling to [8, 256, 256]
Elapsed time (ms): 1.05557
```

Again, this performance slowdown happens only for 3D case, where in the code we unroll a template loop over the dimensions: https://github.com/pytorch/pytorch/compare/master...Quansight:upsample-tensor-iterator-another-dispatch#diff-b924d55052c58e09dfb65f4cbbe17396f0f588eb722cffad5d828ff51462ab52R288


PS: I also updated [full results](https://github.com/vfdev-5/interpolate-tensoriterator/blob/master/step_six/pth_vs_PR_results.log.save) with more tests cases: non contiguous cases.

File: aten/src/ATen/native/UpSample.h [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#pragma once
```
Comment (hameerabbasi) [inline aten/src/ATen/native/UpSample.h:1]: I'd replace this by header guards -- `#pragma once` can have that effect in everything that includes this header.

Comment (hameerabbasi) [reply aten/src/ATen/native/UpSample.h:1]: @vfdev-5 Pointed out to me that everywhere in PyTorch does this.

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp [hunk 2]
```diff
@@ -563,6 +422,280 @@ void upsample_trilinear3d_backward_kernel_impl(
...
+  auto iter = config.build();
+
+  AT_DISPATCH_FLOATING_TYPES(
+      iter.dtype(), "upsample_linearNd", [&] {
+      ti_cpu_upsample_linear<scalar_t, index_t, out_ndims>(iter);
+  });
+
+}
+
+using scale_t = std::vector<c10::optional<double>>;
```
Comment (fmassa) [inline aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp:650]: This seems duplicate from above?

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp [hunk 3]
```diff
@@ -563,6 +422,280 @@ void upsample_trilinear3d_backward_kernel_impl(
...
+    c10::optional<double> scales_w) {
+
+  if (canUse32BitIndexMath(input)) {
+    ti_upsample_linearNd_kernel_impl<int32_t, 3, scale_t>(
+      output, input, align_corners, {scales_d, scales_h, scales_w});
+  } else {
+    ti_upsample_linearNd_kernel_impl<int64_t, 3, scale_t>(
+      output, input, align_corners, {scales_d, scales_h, scales_w});
+  }
+}
```
Comment (fmassa) [inline aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp:697]: Given that those are helper functions, can we maybe refactor this into a single function instead, with the same API as `cpu_upsample_linear`?

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp [hunk 4]
```diff
@@ -563,6 +422,280 @@ void upsample_trilinear3d_backward_kernel_impl(
...
+    c10::optional<double> scales_h,
+    c10::optional<double> scales_w) {
+
+  if (canUse32BitIndexMath(input)) {
+    ti_upsample_linearNd_kernel_impl<int32_t, 2, scale_t>(
+      output, input, align_corners, {scales_h, scales_w});
+  } else {
+    ti_upsample_linearNd_kernel_impl<int64_t, 2, scale_t>(
+      output, input, align_corners, {scales_h, scales_w});  
+  }
```
Comment (fmassa) [inline aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp:664]: Can you comment in the PR the speedups that we get that using `int32_t` instead of always using `int64_t`? This is going to be important to have documented somewhere, as using both in the function will increase the binary size for those functions by 2x.

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp [hunk 5]
```diff
@@ -563,6 +422,280 @@ void upsample_trilinear3d_backward_kernel_impl(
...
+  // input can be NCHW, NCL or NCKHW
+  auto shape = input.sizes().vec();
+  auto strides = input.strides().vec();
+  auto oshape = output.sizes();
+
+  for (int i=0; i<out_ndims; i++) {
+    shape[i + 2] = oshape[i + 2];
+    strides[i + 2] = 0;
+  }
+  auto restrided_input = input.as_strided(shape, strides);
```
Comment (fmassa) [inline aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp:599]: Can we maybe refer to the advanced indexing implementation to describe what's going on here? This can be tricky to understand for someone new to this implementation.

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp [hunk 6]
```diff
@@ -563,6 +422,280 @@ void upsample_trilinear3d_backward_kernel_impl(
   });
 }
 
+// Helper structs and methods for ti_cpu_upsample_linear
+
+// Linear Interpolation structure to compute output value in n-dimensional case.
+// - recursively compute interpolated output for each dimension
```
Comment (fmassa) [inline aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp:428]: Can we say that the interpolation methods that we use are separable, and as such we can compute the interpolation independently per dimension in a recursive way?
Maybe refer to #10482 for more context

**Review by fmassa — COMMENTED** [review]
I've made some minor comments about the structure of the code.

It would be great to understand the slowdown when integrating this implementation into PyTorch for the 3d case.

**Comment by fmassa** [comment]
Additionally, @vfdev-5 can you maybe put the performance results in a table which is linked in the PR? This will make it easier to understand what's going on.

Something like

## 2d
| config | master | this PR |
| --- | --- | --- |
| [1x3x256x256] -> [1x3x500x500] | xxx | xxx |
| [1x3x256x256] -> [1x3x128x128] | xxx | xxx |
...

## 3d
| config | master | this PR |
| --- | --- | --- |
| [1x3x8x256x256] -> [1x3x8x500x500] | xxx | xxx |

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp (hunk 4 above)
Comment (vfdev-5) [reply aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp:664]: Rebenchmarking that I have a doubt now that we have any significant gain when using `int32_t`, I think we can always use `int64_t` and do not worry about 2x binary size etc. What do you think ?

**Comment by vfdev-5** [comment]
Results for 149b976 = if tensor iterator is used instead of cpu_upsample_linear_channels_last.
TensorIterator implementation is faster for some cases and slower for others (see the speedups <1.0):

<details>
<summary>
Interpolation 2d - 6 thread(s)
</summary>


In | Out | Is contigous | Channels last | master | this PR | speed-up
---|---|---|---|---|---|---
[1, 3, 320, 320] | [256, 256] | True | False | 0.3261 | 0.1239 | 2.6318
[1, 3, 320, 320] | [512, 512] | True | False | 1.2854 | 0.4086 | 3.1458
[1, 3, 320, 320] | [256, 256] | False | False | 0.3488 | 0.0777 | 4.4919
[1, 3, 320, 320] | [512, 512] | False | False | 1.3063 | 0.4084 | 3.1984
[1, 3, 320, 320] | [256, 256] | False | True | 1.0897 | 0.3289 | 3.3132
[1, 3, 320, 320] | [512, 512] | False | True | 4.2505 | 1.3841 | 3.0711
[32, 128, 64, 64] | [32, 32] | False | True | 2.2961 | 2.9314 | 0.7833
[32, 128, 64, 64] | [128, 128] | False | True | 35.9384 | 35.6293 | 1.0087
[32, 128, 64, 64] | [32, 32] | True | False | 3.6902 | 3.5451 | 1.0409
[32, 128, 64, 64] | [128, 128] | True | False | 86.7835 | 52.4501 | 1.6546
[1, 3, 500, 500] | [256, 256] | True | False | 0.3266 | 0.0785 | 4.1601
[1, 3, 500, 500] | [800, 800] | True | False | 3.1868 | 0.5580 | 5.7114
[1, 3, 500, 500] | [256, 256] | False | False | 0.3771 | 0.0793 | 4.7555
[1, 3, 500, 500] | [800, 800] | False | False | 3.2693 | 0.5610 | 5.8271


</details>

<details>
<summary>
Interpolation 1d - 6 thread(s)
</summary>


In | Out | Is contigous | Channels last | master | this PR | speed-up
---|---|---|---|---|---|---
[4, 512, 320] | 256 | True | False | 0.2808 | 0.1044 | 2.6907
[4, 512, 320] | 512 | True | False | 0.5524 | 0.1887 | 2.9269


</details>

<details>
<summary>
Interpolation 3d - 6 thread(s)
</summary>


In | Out | Is contigous | Channels last | master | this PR | speed-up
---|---|---|---|---|---|---
[1, 3, 16, 320, 320] | [8, 256, 256] | True | False | 4.4017 | 0.9578 | 4.5958
[1, 3, 16, 320, 320] | [32, 512, 512] | True | False | 84.0302 | 24.0669 | 3.4915
[1, 3, 16, 320, 320] | [8, 256, 256] | False | True | 13.6098 | 3.0288 | 4.4934
[1, 3, 16, 320, 320] | [32, 512, 512] | False | True | 246.6380 | 64.3400 | 3.8334


</details>

<details>
<summary>
Interpolation 2d - 1 thread(s)
</summary>


In | Out | Is contigous | Channels last | master | this PR | speed-up
---|---|---|---|---|---|---
[1, 3, 320, 320] | [256, 256] | True | False | 0.8967 | 0.4551 | 1.9703
[1, 3, 320, 320] | [512, 512] | True | False | 3.5399 | 1.7594 | 2.0120
[1, 3, 320, 320] | [256, 256] | False | False | 0.9760 | 0.3305 | 2.9531
[1, 3, 320, 320] | [512, 512] | False | False | 3.6266 | 1.7643 | 2.0555
[1, 3, 320, 320] | [256, 256] | False | True | 1.0093 | 1.6589 | 0.6084
[1, 3, 320, 320] | [512, 512] | False | True | 4.0231 | 7.1302 | 0.5642
[32, 128, 64, 64] | [32, 32] | False | True | 5.8736 | 9.6382 | 0.6094
[32, 128, 64, 64] | [128, 128] | False | True | 108.2541 | 117.1183 | 0.9243
[32, 128, 64, 64] | [32, 32] | True | False | 19.9122 | 14.0883 | 1.4134
[32, 128, 64, 64] | [128, 128] | True | False | 398.8196 | 205.5317 | 1.9404
[1, 3, 500, 500] | [256, 256] | True | False | 0.8944 | 0.3388 | 2.6404
[1, 3, 500, 500] | [800, 800] | True | False | 8.6327 | 2.9568 | 2.9196
[1, 3, 500, 500] | [256, 256] | False | False | 1.0921 | 0.3405 | 3.2076
[1, 3, 500, 500] | [800, 800] | False | False | 8.9394 | 2.9654 | 3.0145


</details>

<details>
<summary>
Interpolation 1d - 1 thread(s)
</summary>


In | Out | Is contigous | Channels last | master | this PR | speed-up
---|---|---|---|---|---|---
[4, 512, 320] | 256 | True | False | 1.5233 | 0.5066 | 3.0071
[4, 512, 320] | 512 | True | False | 3.0312 | 0.9796 | 3.0943


</details>

<details>
<summary>
Interpolation 3d - 1 thread(s)
</summary>


In | Out | Is contigous | Channels last | master | this PR | speed-up
---|---|---|---|---|---|---
[1, 3, 16, 320, 320] | [8, 256, 256] | True | False | 12.0408 | 4.8498 | 2.4827
[1, 3, 16, 320, 320] | [32, 512, 512] | True | False | 222.8379 | 105.1315 | 2.1196
[1, 3, 16, 320, 320] | [8, 256, 256] | False | True | 13.3036 | 17.2361 | 0.7718
[1, 3, 16, 320, 320] | [32, 512, 512] | False | True | 245.9575 | 297.0317 | 0.8281


</details>


<details>
<summary>
Versions and build configs
</summary>

PyTorch master: 1.8.0.dev20210208+cu110
PyTorch master build setting:
```
BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.0, CUDNN_VERSION=8.0.5, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.8.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON,
```

PR : 1.9.0a0+149b976
PR build setting:
```
BUILD_TYPE=Release, CUDA_VERSION=11.1, CUDNN_VERSION=8.0.5, CXX_COMPILER=/usr/bin/g++-7, CXX_FLAGS=-O3 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.9.0, USE_CUDA=1, USE_CUDNN=1, USE_EIGEN_FOR_BLAS=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=OFF, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=0, USE_OPENMP=ON,
```
</details>

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp [hunk 7]
```diff
@@ -484,15 +491,129 @@ void cpu_upsample_linear_backward(
...
+  output.emplace_back(empty(new_shape, CPU(c10::CppTypeToScalarType<scalar_t>())));  
+  output.emplace_back(empty(new_shape, CPU(at::kLong)));
+  output.emplace_back(empty(new_shape, CPU(c10::CppTypeToScalarType<scalar_t>())));
+
+  auto input_index0_ptr = output[0].data_ptr<int64_t>();
+  auto lambda0_ptr = output[1].data_ptr<scalar_t>();
+  auto input_index1_ptr = output[2].data_ptr<int64_t>();
+  auto lambda1_ptr = output[3].data_ptr<scalar_t>();
+
+  double xd;
```
Comment (fmassa) [inline aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp:528]: This looks unused?

Comment (vfdev-5) [reply aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp:528]: Thanks

**Review by fmassa — APPROVED** [review]
Looks great, thanks a lot!

**Comment by VitalyFedyunin** [comment]
Please do `REGISTER_AVX2_DISPATCH` workaround while I review the issue with dispatch (as it might take additional time).

**Comment by fmassa** [comment]
I propose we do the following: given that the current PR already gives a significant speedup for the 3d case (despite the issues with `REGISTER_DISPATCH`), I would say to merge the PR as is, while keeping in mind that we can still obtain a 2x speed improvement on top of it when we fix the `REGISTER_AVX2_DISPATCH` situation.
